Neilj/improved delegation doc 2 (#4832)
matrix-org/synapse: README.rst

Synapse Installation
====================

> .. _federation:

Reviewer: this doesn't seem to have worked. it may need more whitespace around it or something?

.. _UPGRADE.rst: UPGRADE.rst

> 

Author: Nods

> 

Reviewer: I can't see where you've moved this to?

> 

Author: oversight, added back in.

Using PostgreSQL
================

> Synapse offers two database engines:-

Reviewer: https://github.com/matrix-org/matrix-doc/blob/master/meta/documentation_style.rst#stylistic-notes
```suggestion
Synapse offers two database engines:
```

... and full-stops after the entries.

 * `SQLite <https://sqlite.org/>`_
 * `PostgreSQL <https://www.postgresql.org>`_

By default Synapse uses SQLite in and doing so trades performance for convenience.
SQLite is only recommended in Synapse for testing purposes or for servers with 
light workloads.

> Almost all installations should opt to use PostreSQL, advantages include:-

Reviewer: comma tax.
```suggestion
Almost all installations should opt to use PostreSQL. Advantages include:
```

Building internal API documentation::

    python setup.py build_sphinx

Troubleshooting
===============

Running out of File Handles
---------------------------

> If synapse runs out of filehandles, it typically fails badly - live-locking

Reviewer 2: ```suggestion
If synapse runs out of file handles, it typically fails badly - live-locking
```

at 100% CPU, and/or failing to accept new TCP connections (blocking the
connecting client).  Matrix currently can legitimately use a lot of file handles,
thanks to busy rooms like #matrix:matrix.org containing hundreds of participating
servers.  The first time a server talks in a room it will try to connect
simultaneously to all participating servers, which could exhaust the available
file descriptors between DNS queries & HTTPS sockets, especially if DNS is slow

> to respond.  (We need to improve the routing algorithm used to be better than

Reviewer 2: ```suggestion
to respond. (We need to improve the routing algorithm used to be better than
```

> full mesh, but as of June 2017 this hasn't happened yet).

Reviewer 2: ```suggestion
full mesh, but as of March 2019 this hasn't happened yet).
```

If you hit this failure mode, we recommend increasing the maximum number of
open file handles to be at least 4096 (assuming a default of 1024 or 256).
This is typically done by editing ``/etc/security/limits.conf``

Separately, Synapse may leak file handles if inbound HTTP requests get stuck
during processing - e.g. blocked behind a lock or talking to a remote server etc.
This is best diagnosed by matching up the 'Received request' and 'Processed request'
log lines and looking for any 'Processed request' lines which take more than

> a few seconds to execute.  Please let us know at #synapse:matrix.org if

Reviewer 2: ```suggestion
a few seconds to execute. Please let us know at #synapse:matrix.org if
```

you see this failure mode so we can help debug it, however.

Help!! Synapse eats all my RAM!
-------------------------------

Synapse's architecture is quite RAM hungry currently - we deliberately
cache a lot of recent room data and metadata in RAM in order to speed up

> common requests.  We'll improve this in future, but for now the easiest

Reviewer 2: ```suggestion
common requests. We'll improve this in the future, but for now the easiest
```

Am not a native English speaker, but https://jakubmarian.com/in-the-future-vs-in-future-in-british-and-american-english/ says it's wrong

way to either reduce the RAM usage (at the risk of slowing things down)
is to set the almost-undocumented ``SYNAPSE_CACHE_FACTOR`` environment

> variable.  The default is 0.5, which can be decreased to reduce RAM usage

Reviewer 2: ```suggestion
variable. The default is 0.5, which can be decreased to reduce RAM usage
```
